components/calendar/src/julian.rs

```diff
@@ -0,0 +1,317 @@
+        Self::offset_date(date, offset);
+    }
+
+    #[allow(clippy::field_reassign_with_default)]
+    fn until(
+        &self,
+        date1: &Self::DateInner,
+        date2: &Self::DateInner,
+        largest_unit: DateDurationUnit,
+        smallest_unit: DateDurationUnit,
+    ) -> DateDuration<Self> {
+        Iso.until(&date1.0, &date2.0, largest_unit, smallest_unit)
+            .cast_unit()
+    }
+
+    /// The calendar-specific year represented by `date`
+    fn year(&self, date: &Self::DateInner) -> types::Year {
+        date.0.year.into()
+    }
+
+    /// The calendar-specific month represented by `date`
+    fn month(&self, date: &Self::DateInner) -> types::Month {
+        date.0.month.into()
+    }
+
+    /// The calendar-specific day-of-month represented by `date`
+    fn day_of_month(&self, date: &Self::DateInner) -> types::DayOfMonth {
+        date.0.day.into()
+    }
+
+    fn day_of_year_info(&self, date: &Self::DateInner) -> types::DayOfYearInfo {
+        let prev_year = IsoYear(date.0.year.0 - 1);
+        let next_year = IsoYear(date.0.year.0 + 1);
+        types::DayOfYearInfo {
+            day_of_year: Julian::day_of_year(date.0),
+            days_in_year: Julian::days_in_year(date.0.year),
+            prev_year: prev_year.into(),
+            days_in_prev_year: Julian::days_in_year(prev_year),
+            next_year: next_year.into(),
+        }
+    }
+
+    fn debug_name() -> &'static str {
+        "Julian"
+    }
+}
+
+impl Julian {
+    /// Construct a new Julian Calendar
+    pub fn new() -> Self {
+        Self
+    }
+
+    fn is_leap_year(year: IsoYear) -> bool {
+        year.0 % 4 == 0
+    }
+
+    fn days_in_year(year: IsoYear) -> u32 {
+        if Self::is_leap_year(year) {
+            366
+        } else {
+            365
+        }
+    }
+
+    fn day_of_year(date: IsoDateInner) -> u32 {
+        let month_offset = [0, 1, -1, 0, 0, 1, 1, 2, 3, 3, 4, 4];
+        let mut offset = month_offset[u8::from(date.month) as usize - 1];
+        if Self::is_leap_year(date.year) && u8::from(date.month) > 2 {
+            offset += 1;
+        }
+        let prev_month_days = (30 * (u8::from(date.month) as i32 - 1) + offset) as u32;
+        prev_month_days + u8::from(date.day) as u32
+    }
+
+    fn days_in_month(year: IsoYear, month: IsoMonth) -> u8 {
+        match u8::from(month) {
+            4 | 6 | 9 | 11 => 30,
+            2 if Self::is_leap_year(year) => 29,
+            2 => 28,
+            _ => 31,
+        }
+    }
+
+    fn calculate_day_difference_between_calendars(date: IsoDateInner) -> i32 {
+        // In year 0, the slack is -2
+        let year = date.year.0;
+        let slack = year / 100 - year / 400 - 2;
+
+        if year % 100 == 0 && year % 400 != 0 && u8::from(date.month) <= 2 {
+            slack - 1
+        } else {
+            slack
+        }
+    }
+
+    fn offset_date(date: &mut JulianDateInner, mut offset: DateDuration<Self>) -> JulianDateInner {
```

Comment:
thought: I'd really like this code to be refactored out and made common, but perhaps I can do this for you once the code lands? This will be useful for the indian national calendar as well

Reply:
Maybe I can pick up the refactoring out as well if we know for sure where the helper methods are supposed to be.